plugins/inputs/dpdk/dpdk_connector.go

```diff
@@ -0,0 +1,145 @@
+// +build linux
+
+package dpdk
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+)
+
+const maxInitMessageLength = 1024
+
+type initMessage struct {
+	Version      string `json:"version"`
+	Pid          int    `json:"pid"`
+	MaxOutputLen uint32 `json:"max_output_len"`
+}
+
+type dpdkConnector struct {
+	pathToSocket  string
+	maxOutputLen  uint32
+	messageShowed bool
+	accessTimeout time.Duration
+	connection    net.Conn
+}
+
+func newDpdkConnector(pathToSocket string, accessTimeout config.Duration) *dpdkConnector {
+	return &dpdkConnector{
+		pathToSocket:  pathToSocket,
+		messageShowed: false,
+		accessTimeout: time.Duration(accessTimeout),
+	}
+}
+
+// Connects to the socket
+// Since DPDK is a local unix socket, it is instantly returns error or connection, so there's no need to set timeout for it
+func (conn *dpdkConnector) connect() (string, error) {
+	connection, err := net.Dial("unixpacket", conn.pathToSocket)
+
+	if err != nil {
+		return "", fmt.Errorf("failed to connect to the socket - %v", err)
+	}
+
+	result, err := conn.readMaxOutputLen(connection)
+	if err != nil {
+		return result, err
+	}
+
+	conn.connection = connection
+	return result, nil
+}
+
+// Executes command using provided connection and returns response
+// If error (such as timeout) occurred, then connection is discarded and recreated
+// because otherwise behaviour of connection is undefined (e.g. it could return result of timed out command instead of latest)
+func (conn *dpdkConnector) getCommandResponse(fullCommand string) ([]byte, error) {
+	connection, err := conn.getConnection()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get connection to execute %v command - %v", fullCommand, err)
+	}
+
+	err = conn.setTimeout(connection)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set timeout for %v command - %v", fullCommand, err)
+	}
+
+	_, err = connection.Write([]byte(fullCommand))
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("send '%v'", fullCommand), err)
+	}
+
+	buf := make([]byte, conn.maxOutputLen)
+	messageLength, err := connection.Read(buf)
+	if err != nil {
+		return nil, conn.tryClose(fmt.Sprintf("read response of '%v'", fullCommand), err)
+	}
+
+	if messageLength == 0 {
+		return nil, fmt.Errorf("got empty response during execution of '%v' command", fullCommand)
+	}
+	return buf[:messageLength], nil
+}
+
+func (conn *dpdkConnector) tryClose(operation string, err error) error {
+	closeErr := conn.connection.Close()
+	conn.connection = nil
+	newErr := fmt.Sprintf("failed to %v command - %v", operation, err)
```

Comment:
Wow this is so wrong. Passing down an `err` here just to reformat it and return it home!? Can we please handle this at the level where this function is called!

Reply:
Handled.